src/databricks/labs/ucx/azure/access.py

```diff
@@ -79,27 +87,110 @@ def save_spn_permissions(self) -> str | None:
             )
             return None
         storage_account_infos = []
-        for storage in self._azurerm.storage_accounts():
-            if storage.storage_account not in used_storage_accounts:
-                continue
+        for storage in used_storage_accounts:
             for mapping in self._map_storage(storage):
                 storage_account_infos.append(mapping)
         if len(storage_account_infos) == 0:
             logger.error("No storage account found in current tenant with spn permission")
             return None
         return self._installation.save(storage_account_infos, filename=self._filename)
 
+    def _update_cluster_policy_definition(
+        self, policy_definition: str, storage_accounts: list[AzureResource], global_principal: Principal
+    ) -> str:
+        policy_dict = json.loads(policy_definition)
+        tenant_id = self._azurerm.tenant_id()
+        endpoint = f"https://login.microsoftonline.com/{tenant_id}/oauth2/token"
+        for storage in storage_accounts:
+            policy_dict[
+                f"spark_conf.fs.azure.account.oauth2.client.id.{storage.storage_account}.dfs.core.windows.net"
+            ] = self._policy_config(global_principal.client_id)
+            policy_dict[
+                f"spark_conf.fs.azure.account.oauth.provider.type.{storage.storage_account}.dfs.core.windows.net"
+            ] = self._policy_config("org.apache.hadoop.fs.azurebfs.oauth2.ClientCredsTokenProvider")
+            policy_dict[
+                f"spark_conf.fs.azure.account.oauth2.client.endpoint.{storage.storage_account}.dfs.core.windows.net"
+            ] = self._policy_config(endpoint)
+            policy_dict[f"spark_conf.fs.azure.account.auth.type.{storage.storage_account}.dfs.core.windows.net"] = (
+                self._policy_config("OAuth")
+            )
+            if global_principal.secret is not None:
+                policy_dict[
+                    f"spark_conf.fs.azure.account.oauth2.client.secret.{storage.storage_account}.dfs.core.windows.net"
+                ] = self._policy_config(global_principal.secret)
+        return json.dumps(policy_dict)
+
+    @staticmethod
+    def _policy_config(value: str):
+        return {"type": "fixed", "value": value}
+
+    def _update_cluster_policy_with_spn(
+        self, policy_id: str, storage_accounts: list[AzureResource], global_principal: Principal
+    ):
+        try:
+            policy_definition = ""
+            cluster_policy = self._ws.cluster_policies.get(policy_id)
+            self._installation.save(cluster_policy.as_dict(), filename="policy-backup.json")
```

Comment:
```suggestion
            self._installation.save(cluster_policy, filename="policy-backup.json")
```
No need for as_dict()